framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java

```diff
@@ -497,14 +505,99 @@ public void typeProcess(TypeElement element, TreePath tree) {
+            warnUnneededSuppressions();
+        }
         if (getSubcheckers().size() > 0) {
             printCollectedMessages(tree.getCompilationUnit());
             // Update errsOnLastExit to reflect the errors issued.
             this.errsOnLastExit = log.nerrors;
         }
     }
 
+    /**
+     * Issues a warning about any {@code @SuppressWarnings} that isn't used by this checker, but
+     * contains a key that would suppress a warning from this checker.
+     */
+    private void warnUnneededSuppressions() {
+        if (!hasOption("warnUnneededSuppressions")) {
+            return;
+        }
+        Set<Element> elementsSuppress = new HashSet<>(this.elementsWithSuppressedWarnings);
+        this.elementsWithSuppressedWarnings.clear();
+        Set<String> checkerKeys = new HashSet<>(getSuppressWarningsKeys());
+        Set<String> errorKeys = new HashSet<>(messages.stringPropertyNames());
+        for (BaseTypeChecker subChecker : subcheckers) {
+            elementsSuppress.addAll(subChecker.elementsWithSuppressedWarnings);
+            subChecker.elementsWithSuppressedWarnings.clear();
+            checkerKeys.addAll(subChecker.getSuppressWarningsKeys());
+            errorKeys.addAll(subChecker.messages.stringPropertyNames());
+            subChecker.getVisitor().treesWithSuppressWarnings.clear();
+        }
+        // It's not clear for which checker this suppression is intended,
+        // so never report it as unused.
+        checkerKeys.remove(SourceChecker.SUPPRESS_ALL_KEY);
+
+        for (Tree tree : getVisitor().treesWithSuppressWarnings) {
+            Element elt = TreeUtils.elementFromTree(tree);
+            SuppressWarnings suppressAnno = elt.getAnnotation(SuppressWarnings.class);
+            if (suppressAnno == null || elementsSuppress.contains(elt)) {
+                continue;
+            }
+            for (String keyFromAnno : suppressAnno.value()) {
+                for (String checkerKey : checkerKeys) {
+                    if (keyFromAnno.contains(checkerKey)) {
+                        Tree swTree = findSWTree(tree);
+                        report(
+                                Result.warning(
+                                        SourceChecker.UNNEEDED_SUPPRESSION_KEY,
+                                        getClass().getSimpleName(),
+                                        "\"" + keyFromAnno + "\""),
+                                swTree);
+                    }
+                }
+                if (keyFromAnno.contains(":")) {
+                    // The key starts with a checker name, if that is this checker, then the warning
+                    // was issued above.  For example, if this is the Nullness Checker and the
+                    // keyForAnno is "index:override.return.invalid", then don't issue a warning.
+                    continue;
+                }
+
+                for (String errorKey : errorKeys) {
+                    if (keyFromAnno.equals(errorKey)) {
```

Comment:
Some documentation would be helpful to understand what these three nested for loops each do.
Why can the `checkerKeys` and `errorKeys` lists not be merged? Why is one using `contains` and the other `equals`?